### PR TITLE
feat: Parallelise Model Loading

### PR DESCRIPTION
**Pull request by vovw**
test using 
```
exo --preload-models llama-3.2-1b,llama-3.1-8b              
```

### CODE REVIEW

**Comment by AlexCheema** [comment]
Almost what I envisioned - only thing I would change is to preload after the preemptive download. We don't want to download all possible model shards, only the relevant one.

File: exo/main.py [hunk 1]
```diff
@@ -100,16 +101,18 @@
...
-  except Exception as e:
-    if DEBUG >= 2:
-      print(f"Failed to preemptively start download: {e}")
-      traceback.print_exc()
+    try:
+        status = json.loads(opaque_status)
+        if status.get("type") == "node_status" and status.get("status") == "start_process_prompt":
+            current_shard = node.get_current_shard(Shard.from_dict(status.get("shard")))
+            if DEBUG >= 2: print(f"Preemptively starting download for {current_shard}")
+            asyncio.create_task(shard_downloader.ensure_shard(current_shard))
```
Comment (AlexCheema) [inline exo/main.py:109]: I was thinking we preload here, after `shard_downloader.ensure_shard(current_shard)`

**Review by AlexCheema — CHANGES_REQUESTED** [review]

File: exo/main.py (hunk 1 above)
Comment (vovw) [reply exo/main.py]: just resolved this, can you check if I am missing something else?

**Comment by vovw** [comment]
tested on a m3 pro

File: exo/main.py [hunk 2]
```diff
@@ -195,6 +203,18 @@ def handle_exit():
...
 
+  if args.preload_models:
+    models_to_preload = [Shard.from_dict(model_base_shards[model_name][inference_engine.__class__.__name__]) 
+                         for model_name in args.preload_models.split(',')]
+    for shard in models_to_preload:
+      current_shard = preemptively_start_download(str(uuid.uuid4()), json.dumps({
+        "type": "node_status",
+        "status": "start_process_prompt",
+        "shard": shard.to_dict()
+      }))
```
Comment (AlexCheema) [inline exo/main.py:214]: doesn't this need to be awaited?

File: exo/main.py [hunk 3]
```diff
@@ -195,6 +203,18 @@ def handle_exit():
...
+      if current_shard:
+        await node.preload_models([current_shard])
```
Comment (AlexCheema) [inline exo/main.py:216]: doesn't `preemptively_start_download` already call `preload_models`?

**Review by AlexCheema — CHANGES_REQUESTED** [review]

**Comment by vovw** [comment]
closed this one because it has been >month for the last pr, synced my fork and the new pr is at #466

**Pull request closed without merging**